text/0000-extern-types.md

```diff
@@ -0,0 +1,110 @@
+- Feature Name: extern_types
+- Start Date: 2017-01-18
+- RFC PR: 
+- Rust Issue: 
+
+# Summary
+[summary]: #summary
+
+Add an `extern type` syntax for declaring types which are opaque to Rust's type
+system.
+
+# Motivation
+[motivation]: #motivation
+
+When interacting with external libraries we often need to be able to handle
+pointers to data that we don't know the size or layout of.
+
+In C it's possible to declare a type but not define it. These incomplete types
+can only be used behind pointers, a compilation error will result if the
+user tries to use them in such a way that the compiler would need to know their
+layout.
+
+In Rust, we don't have this feature. Instead, a couple of problematic hacks are
+used in its place.
+
+One is, we define the type as an uninhabited type. eg.
+
+    enum MyFfiType {}
+
+Another is, we define the type with a private field and no methods to construct
+it.
+
+    struct MyFfiType {
+        _priv: (),
+    }
+
+The point of both these constructions is to prevent the user from being able to
+create or deal directly with instances of
+the type. Niether of these types accurately reflect the reality of the
+situation. The first definition is logically problematic as it defines a type
+which can never exist. This means that references to the type can also -
+logically - never exist and raw pointers to the type are guaranteed to be
+invalid. The second definition says that the type is a ZST, that we can store
+it on the stack and that we can call `ptr::read`, `mem::size_of` etc. on it.
+None of this is of course valid.
+
+This RFC instead proposes a way to directly express that a type exists but is
+unknown to Rust.
+
+# Detailed design
+[design]: #detailed-design
+
+Add a new kind of type declaration, an `extern type`:
+
+    extern type Foo;
```

Comment:
This diverges somewhat from `extern fn` where its a *definition* and not declaration. For consistency IMHO declarations should stay constrained to `extern` blocks, like below.